Documentation/proposals/202304-silences.md

```diff
@@ -0,0 +1,113 @@
+  to fully manage the life cycle of Silences Custom resources, Nevertheless, this comes with a drawback
+  as the team must be able to manage a new component in their stack.
+
+Additionally, Having a new component in the stack and keeping it maintained is not always ideal (said the folks at [Giant Swarm](https://giantswarm.io) the owners of [silence-operator](https://github.com/giantswarm/silence-operator)).
+It's really better to have this feature as part of prometheus-operator and keep it maintained and available for the community.
+
+By Adding support for `Silence` CRD in the prometheus-operator, this will make users more flexible in terms of choosing the tool
+to deploy their silences and will free others from managing a standalone component within the stack.
+
+## Pitfalls of the current solution
+
+Using Alertmanager API Directly comes with drawbacks:
+
+* Teams have to build an automation to add silences in a centralized manner
+* There is no input validation, which can lead to an invalid silence configuration
+* Teams need to manage expiration on their own
+* If a silence was deleted by accident, it will be permanent.
+
+# Goals
+
+* Provide a way for users to manage Alertmanager silences with the Kubernetes Way.
+* Make it easier to manage Alertmanager silences via centralised repo with e.g. ArgoCD, Flux...
+* Provide a way to manage silence expiration with a silence CR.
+
+## Audience
+
+* Users who serve Prometheus as a service and want to have an interface in defining silences exposed to developers.
+* Users who want to manage silences the same way as for services running within the Kubernetes cluster
+* Users who want a supported Kubernetes way of silences outside the Kubernetes cluster
+
+# Non-Goals
+
+* This proposal doesn't aim to remove the CR after the expiration date has reached
+* Refactoring of the other CRDs is not in scope for the first version
+
+# How
+
+Creating a new cluster-scope Silence CRD that will act as an interface by adding silences via the Alertmanager API.
+Usage of Silences doesn't exclude the use of the other CRDs, they are not mutually exclusive.
+`Silence` will allow the creation of any silences to Alertmanager, while the other CRDs provide sane defaults. This will allow
+for isolated testing of the new `Silence` CRD.
+
+A typical Silence resouce should looks like the following:
+
+```yaml
+apiVersion: monitoring.coreos.com/v1alpha1
+kind: Silence
+metadata:
+  name: my-silence
+  labels:
+    test: value
+spec:
+  selector:
+    app: test-alertmanager
+  validUntil: "2023-06-01"
+  matchers:
+    - name: namespace
+      value: test-ns
+      isRegex: false
+```
+
+The above resource will result in creating a silence in the Alertmanager with label selector,
+`app: test-alertmanager`.
+
+* `selector` used to select which Kubernetes Service to use as an Alertmanager API.
+* `validUntil` to define the expiration of the silence.
```

Comment:
Would it make sense to have silences that stay active so long as the resource exists? For instance, if `validUntil` is undefined, then the Silence stays active. If it is, the `Silence` expires at the date defined.

Reply:
Yeah, I'm guessing if we can add a default maximum days if the user has not defined the `validUntil `

Reply:
we can make the field `expiresAt` required so :)

Reply:
The implementation could also keep the silence active when this field is not defined. I have no strong opinion on this.

Reply:
Got it, so:
* If a silence has `expiredAt`, it'll just expire at the time defined and the CR will have to be manually removed
* If a silence doesn't have `expiredAt`, it'll be kept active as long as the CR exists and removed as soon as the CR is gone

I'm not sure how do-able the second point is but I think it'd make sense, especially with GitOps.

Is that what we aim for?

Reply:
I think so,

Reply:
However,  how do we handle silence overwrite?

Reply:
Silence operator would update them. The silence id in alertmanager would be kept in the status

Reply:
Great! Can you add that to the proposal? I guess in the How section.

Reply:
Added!